### PR TITLE
Revision hough circles

### PR DESCRIPTION
**Pull request by mschoeneck**
Restructuring HoughCircles using more C++ like coding style. Adding SIMD code where useful and parallelizing the for loops.
I have also enabled the maximum number of circles as a pre-set parameter. If it is set to zero, no radius will be calculated, but a sorted list of detected centers is returned. If the value is negative the maximum number is INT_MAX.

Please test my new implementation for HoughCircles. For comparing the old code is used when cvHoughCircles is called.

I tried to use universal intrinsics for SIMD code, but currently I don't find a good solution in the HoughCircleEstimateRadiusInvoker class.

The detection is stable, but multithreading not always gives a good performance gain. When only a single thread is used some overheads of multithreading is avoided.

I also optimized the last sorting, which should give a good performance gain compared to the old implementation. So even single threaded it should be faster.

Further acceleration can be reached with a faster sorting routine and a thread safe container for the accumulation image.

When everything is fine I will remove the old code.

### CODE REVIEW

**Comment by mschoeneck** [comment]
It seems that the extra parameter in HoughCircles causes the error in Compare ABI dumps. In my point of view this should not lead to incompatibilities because of the pre-set of this parameter. So what can I do to avoid this error?

**Comment by alalek** [comment]
You need to restore an old function with implementation which just calls the new function. Parameter defaults should be removed from extended version to resolve ambiguous all error.

**Comment by mschoeneck** [comment]
Why do I always get errors in modules my changes don't affect such as stitching test or contrib docs? What shall I do?

**Comment by vpisarev** [comment]
@mschoeneck, thank you for the PR! That's a massive patch! What's the performance increase you've got with this patch?

**Comment by mschoeneck** [comment]
Performance gain is a bit difficult, depending on image and number of circles. I think the problem is that there is no thread safe container like in TBB library (For accumulate Hough image).

For just a few circles, there is not really a performance gain (In my test case (1 circle): ~45ms -> ~40ms), whereas for a lot of circles I get a good performance scale with number of threads especially in the last step (unrealistic test case (minDist=1): ~1700ms -> ~330ms with 4 cores, single thread I get ~1300ms because of more efficient sort function compared to old implementation). 

I think the new code is a good base for future work, if there will be thread safe containers available. In principle I got a general increase of performance because of sort function and separated Sobel and Canny function. In a lot of situations the accumulation of Hough image is the bottle neck, but at the moment there is no good solution to speed this up.

File: modules/imgproc/src/hough.cpp [hunk 1]
```diff
@@ -1002,222 +1006,770 @@ cvHoughLines2( CvArr* src_image, void* lineStorage, int method,
...
 \****************************************************************************************/
 
-static void
-icvHoughCirclesGradient( CvMat* img, float dp, float min_dist,
-                         int min_radius, int max_radius,
-                         int canny_threshold, int acc_threshold,
-                         CvSeq* circles, int circles_max )
+#if CV_SIMD128
+static
+inline unsigned int trailingZeros(unsigned int value) {
```
Comment (alalek) [inline modules/imgproc/src/hough.cpp]: We have the same function in contours.cpp.
It is very generic function, so it is better to move these implementations into [opencv2/core/hal/intrin.hpp](https://github.com/opencv/opencv/blob/master/modules/core/include/opencv2/core/hal/intrin.hpp) file (at the end of file).
Probably under this signature (and without `#if CV_SIMD128`):
```
inline unsigned int trailingZeros32(unsigned int value)
```

**Comment by mschoeneck** [comment]
I wonder how the planning for C++11 is. It seems to be OK if I declare a pointer in that way to get a thread safe Mat access:
`std::atomic<int> *adata = accum.ptr<std::atomic<int>>(); `
I didn't make performance tests until now, but it builds in Visual Studio and GCC and seems to work as expected.
Perhabs it could be checked using a define in that kind:
```
#if __cpluplus >= 201103L
#define HAVE_CPP11
#endif
```
Is there any experience with that?

**Comment by savuor** [comment]
This PR was fixed and reopened as PR #10041

**Pull request closed without merging**